src/cls/cmpomap/server.cc

```diff
@@ -0,0 +1,294 @@
+ * Foundation.  See file COPYING.
+ */
+
+#include "objclass/objclass.h"
+#include "ops.h"
+
+CLS_VER(1,0)
+CLS_NAME(cmpomap)
+
+using namespace cls::cmpomap;
+
+// returns negative error codes or 0/1 for failed/successful comparisons
+template <typename T>
+static int compare_values(Op op, const T& lhs, const T& rhs)
+{
+  switch (op) {
+  case Op::EQ:  return (lhs == rhs);
+  case Op::NE:  return (lhs != rhs);
+  case Op::GT:  return (lhs > rhs);
+  case Op::GTE: return (lhs >= rhs);
+  case Op::LT:  return (lhs < rhs);
+  case Op::LTE: return (lhs <= rhs);
+  default:      return -EINVAL;
+  }
+}
+
+static int compare_values_u64(const Comparison& cmp, uint64_t lhs,
+                              const bufferlist& value)
+{
+  // an empty existing value defaults to 0 for comparisons
+  uint64_t rhs = 0;
+  if (value.length()) {
+    try {
+      // decode existing value as rhs
+      auto p = value.cbegin();
+      using ceph::decode;
+      decode(rhs, p);
+    } catch (const buffer::error&) {
+      // failures to decode existing values are reported as EIO
+      return -EIO;
+    }
+  }
+  return compare_values(cmp.op, lhs, rhs);
+}
+
+static int compare_value(const Comparison& cmp, const bufferlist& value)
+{
+  switch (cmp.mode) {
+  case Mode::String:
+    return compare_values(cmp.op, cmp.value, value);
+  case Mode::U64:
+    try {
+      // decode input value as lhs
+      uint64_t lhs;
+      auto p = cmp.value.cbegin();
+      using ceph::decode;
+      decode(lhs, p);
+      return compare_values_u64(cmp, lhs, value);
+    } catch (const buffer::error&) {
+      // failures to decode input values are reported as EINVAL
+      return -EINVAL;
+    }
+  default:
+    return -EINVAL;
+  }
+}
+
+static int cmp_vals(cls_method_context_t hctx, bufferlist *in, bufferlist *out)
+{
+  cmp_vals_op op;
+  try {
+    auto p = in->cbegin();
+    decode(op, p);
+  } catch (const buffer::error&) {
+    CLS_LOG(1, "ERROR: cmp_vals(): failed to decode input");
+    return -EINVAL;
+  }
+
+  // collect the keys we need to read
+  std::set<std::string> keys;
+  for (const auto& kv : op.entries) {
+    keys.insert(kv.first);
+  }
+
+  // read the values for each key to compare
+  std::map<std::string, bufferlist> values;
+  int r = cls_cxx_map_get_vals_by_keys(hctx, keys, &values);
+  if (r < 0) {
+    CLS_LOG(4, "ERROR: cmp_vals() failed to read values r=%d", r);
+    return r;
+  }
+
+  auto v = values.cbegin();
+  for (const auto& [key, cmp] : op.entries) {
+    // if the key doesn't exist, use an empty value for comparisons
+    bufferlist value;
+    if (v != values.end() && v->first == key) {
+      value = std::move(v->second);
+      ++v;
+      CLS_LOG(20, "cmp_vals() comparing key=%s op=%d mode=%d",
+              key.c_str(), (int)cmp.op, (int)cmp.mode);
+    } else {
+      CLS_LOG(20, "cmp_vals() comparing missing key=%s op=%d mode=%d",
+              key.c_str(), (int)cmp.op, (int)cmp.mode);
+    }
+
+    r = compare_value(cmp, value);
```

Comment:
why do you bother comparing missing keys?

Reply:
the use case for multisite is to store the most recent timestamps, so it calls cmp_set_vals() with Op::GT. if keys don't exist, i want them to compare less than the input so the input timestamps get written. i think it's useful for cmp_vals() to make similar assertions

Reply:
in 3-value logic, any comparison against "null" should be "null", which eventually should propagate to "false". e.g 
```
5 < null -> null -> false
5 >= null -> null -> false
```
this would work for your example if Op::GT would evaluate t0 "false". but really not depended with the values or operator.

Reply:
if all comparisons against missing keys returned false, then `cmp_set_vals()` would never be able to create keys. this behavior with respect to missing keys is documented with the Ops:
```c++
/// comparison operation, where the left-hand operand is the input value and
/// the right-hand operand is the stored value (or 0/empty if missing)
enum class Op : uint8_t {
```

Reply:
I'm worried about consistency and surprises to the user.
maybe better to be add a parameter to the op so our null behavior is more explicit?
is treating null as zero the existing comparison logic?